// terriajs/dist/Traits/TraitsClasses/ArcGisFeatureServerCatalogItemTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import primitiveArrayTrait from "../Decorators/primitiveArrayTrait";
import primitiveTrait from "../Decorators/primitiveTrait";
import { traitClass } from "../Trait";
import mixTraits from "../mixTraits";
import CatalogMemberTraits from "./CatalogMemberTraits";
import FeaturePickingTraits from "./FeaturePickingTraits";
import { GeoJsonTraits } from "./GeoJsonTraits";
import LegendOwnerTraits from "./LegendOwnerTraits";
import MappableTraits from "./MappableTraits";
import { MinMaxLevelTraits } from "./MinMaxLevelTraits";
import UrlTraits from "./UrlTraits";
let ArcGisFeatureServerCatalogItemTraits = class ArcGisFeatureServerCatalogItemTraits extends mixTraits(UrlTraits, MappableTraits, CatalogMemberTraits, LegendOwnerTraits, GeoJsonTraits, FeaturePickingTraits, MinMaxLevelTraits) {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "clampToGround", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: true
        });
        Object.defineProperty(this, "useStyleInformationFromService", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: true
        });
        Object.defineProperty(this, "layerDef", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "1=1"
        });
        Object.defineProperty(this, "where", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "1=1"
        });
        Object.defineProperty(this, "maxFeatures", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 5000
        });
        Object.defineProperty(this, "featuresPerRequest", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 1000
        });
        Object.defineProperty(this, "supportsPagination", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "tileRequests", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: true
        });
        Object.defineProperty(this, "clipToRectangle", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: true
        });
        Object.defineProperty(this, "maxTiledFeatures", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 100000
        });
        Object.defineProperty(this, "featuresPerTileRequest", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 4000
        });
        Object.defineProperty(this, "maxRecordCountFactor", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 1
        });
        Object.defineProperty(this, "supportsQuantization", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: true
        });
        Object.defineProperty(this, "objectIdField", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "OBJECTID"
        });
        Object.defineProperty(this, "outFields", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: ["OBJECTID"]
        });
    }
};
__decorate([
    primitiveTrait({
        type: "boolean",
        name: "Clamp to Ground",
        description: "Whether the features in this service should be clamped to the terrain surface."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "clampToGround", void 0);
__decorate([
    primitiveTrait({
        type: "boolean",
        name: "Use style information from service",
        description: "Whether to symbolise the data using the drawingInfo object available in the service endpoint."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "useStyleInformationFromService", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "layerDef",
        description: "DEPRECATED, use `where` instead. The 'layerDef' string to pass to the server when requesting geometry."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "layerDef", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "Where clause",
        description: "The 'where' string to pass to the server when requesting geometry."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "where", void 0);
__decorate([
    primitiveTrait({
        type: "number",
        name: "Maximum features",
        description: "The maximum number of features to be retrieved from the feature service."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "maxFeatures", void 0);
__decorate([
    primitiveTrait({
        type: "number",
        name: "Features per request",
        description: "The number of features to be retrieved from the feature service in each request. This should be equal to the " +
            "maxRecordCount specified by the server."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "featuresPerRequest", void 0);
__decorate([
    primitiveTrait({
        type: "boolean",
        name: "Supports pagination",
        description: "Whether this feature service supports pagination. By default, this will be inferred from the server's response."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "supportsPagination", void 0);
__decorate([
    primitiveTrait({
        type: "boolean",
        name: "Tile requests",
        description: "Whether this feature service supports tiled requests. This will be true by default, if the server supports tiling and no unsupported Point/Label/Trail styles are used (for example custom marker images)"
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "tileRequests", void 0);
__decorate([
    primitiveTrait({
        type: "boolean",
        name: "Clip tile requests to rectangle",
        description: `Gets or sets a value indicating whether this dataset should be clipped to the {@link CatalogItem#rectangle}.
If true, no part of the dataset will be displayed outside the rectangle.
This property is true by default, leading to better performance and avoiding tile request errors that might occur when requesting tiles outside the server-specified rectangle.
However, it may also cause features to be cut off in some cases, such as if a server reports an extent that does not take into account that the representation of features sometimes require a larger spatial extent than the features themselves.
For example, if a point feature on the edge of the extent is drawn as a circle with a radius of 5 pixels, half of that circle will be cut off.`
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "clipToRectangle", void 0);
__decorate([
    primitiveTrait({
        type: "number",
        name: "Maximum features",
        description: "For each tile, the maximum number of features to be retrieved from the feature service. This will limit the number of requests, each request uses featuresPerTileRequest trait."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "maxTiledFeatures", void 0);
__decorate([
    primitiveTrait({
        type: "number",
        name: "Features per request",
        description: "The number of features to be retrieved from the feature service in each request. This should be equal to the " +
            "tileMaxRecordCount specified by the server."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "featuresPerTileRequest", void 0);
__decorate([
    primitiveTrait({
        type: "number",
        name: "maxRecordCountFactor",
        description: "When set, the maximum number of features returned by the query will equal the maxRecordCount of the service multiplied by this factor. This only applies to tiled requests. This will use the server's default value."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "maxRecordCountFactor", void 0);
__decorate([
    primitiveTrait({
        type: "boolean",
        name: "Tile requests",
        description: "Whether this feature service supports tiled requests. By default, this will be inferred from the server's response."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "supportsQuantization", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "Object ID Field",
        description: "The field in the feature service that contains the unique identifier for each feature."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "objectIdField", void 0);
__decorate([
    primitiveArrayTrait({
        type: "string",
        name: "Out Fields",
        description: "The fields to be included in the response from the feature service. This will default to the object ID field, and include any fields required for styling. Currently, this only applies to tiled requests."
    })
], ArcGisFeatureServerCatalogItemTraits.prototype, "outFields", void 0);
ArcGisFeatureServerCatalogItemTraits = __decorate([
    traitClass({
        description: `Creates a single item in the catalog from one ESRI Feature Server layer.

  <strong>Note:</strong> <i>Must specify <b>layer ID</b>, e.g. <code>/0</code>, in the URL path.</i>

  Some traits are only relevant when the service supports tiling (see \`tileRequests\` trait).`,
        example: {
            url: "https://services5.arcgis.com/OvOcYIrJnM97ABBA/arcgis/rest/services/Australian_Public_Hospitals_WFL1/FeatureServer/0",
            type: "esri-featureServer",
            name: "Australian Public Hospitals",
            id: "some id"
        }
    })
], ArcGisFeatureServerCatalogItemTraits);
export default ArcGisFeatureServerCatalogItemTraits;
