// terriajs/dist/Traits/TraitsClasses/ArcGisImageServerCatalogItemTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import anyTrait from "../Decorators/anyTrait";
import objectArrayTrait from "../Decorators/objectArrayTrait";
import objectTrait from "../Decorators/objectTrait";
import primitiveArrayTrait from "../Decorators/primitiveArrayTrait";
import primitiveTrait from "../Decorators/primitiveTrait";
import ModelTraits from "../ModelTraits";
import { traitClass } from "../Trait";
import mixTraits from "../mixTraits";
import CatalogMemberTraits from "./CatalogMemberTraits";
import DiscretelyTimeVaryingTraits from "./DiscretelyTimeVaryingTraits";
import ImageryProviderTraits from "./ImageryProviderTraits";
import LayerOrderingTraits from "./LayerOrderingTraits";
import LegendOwnerTraits from "./LegendOwnerTraits";
import { MinMaxLevelTraits } from "./MinMaxLevelTraits";
import UrlTraits from "./UrlTraits";
export class ArcGisImageServerRenderingRule extends ModelTraits {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "rasterFunction", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "variableName", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "rasterFunctionArguments", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
    }
}
__decorate([
    primitiveTrait({
        type: "string",
        name: "Name",
        description: "The name of the raster function."
    })
], ArcGisImageServerRenderingRule.prototype, "rasterFunction", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "Description",
        description: 'optional for well known functions, default is "Raster" for raster function templates.'
    })
], ArcGisImageServerRenderingRule.prototype, "variableName", void 0);
__decorate([
    anyTrait({
        name: "Arguments",
        description: "Overwrite the raster function default configuration by specifying argument parameters (argument names and value types are defined by the author of each raster function template and are not discoverable through ArcGis REST API)."
    })
], ArcGisImageServerRenderingRule.prototype, "rasterFunctionArguments", void 0);
export class ArcGisImageServerAvailableRasterFunctionTraits extends ModelTraits {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "description", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "help", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
    }
}
__decorate([
    primitiveTrait({
        type: "string",
        name: "Name",
        description: "The name of the raster function."
    })
], ArcGisImageServerAvailableRasterFunctionTraits.prototype, "name", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "Description",
        description: "The description of the raster function."
    })
], ArcGisImageServerAvailableRasterFunctionTraits.prototype, "description", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "Unit",
        description: "Help text for the raster function"
    })
], ArcGisImageServerAvailableRasterFunctionTraits.prototype, "help", void 0);
let ArcGisImageServerCatalogItemTraits = class ArcGisImageServerCatalogItemTraits extends mixTraits(ImageryProviderTraits, LayerOrderingTraits, UrlTraits, CatalogMemberTraits, LegendOwnerTraits, DiscretelyTimeVaryingTraits, MinMaxLevelTraits) {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "maximumScale", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "parameters", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "tokenUrl", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "token", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "maxRefreshIntervals", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 10000
        });
        Object.defineProperty(this, "wkid", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 102100
        });
        Object.defineProperty(this, "usePreCachedTiles", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "bandIds", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "renderingRule", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "allowRasterFunction", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "availableRasterFunctions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "disableRasterFunctionSelectors", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
    }
};
__decorate([
    primitiveTrait({
        type: "number",
        name: "Maximum scale",
        description: "Gets or sets the denominator of the largest scale (smallest denominator) for which tiles should be requested.  For example, if this value is 1000, then tiles representing a scale larger than 1:1000 (i.e. numerically smaller denominator, when zooming in closer) will not be requested.  Instead, tiles of the largest-available scale, as specified by this property, will be used and will simply get blurier as the user zooms in closer. Note: maximumLevel overrides this property."
    })
], ArcGisImageServerCatalogItemTraits.prototype, "maximumScale", void 0);
__decorate([
    anyTrait({
        name: "Parameters",
        description: "Additional parameters to pass to the ImageServer when requesting images."
    })
], ArcGisImageServerCatalogItemTraits.prototype, "parameters", void 0);
__decorate([
    primitiveTrait({
        name: "Token URL",
        description: "URL to use for fetching request tokens",
        type: "string"
    })
], ArcGisImageServerCatalogItemTraits.prototype, "tokenUrl", void 0);
__decorate([
    primitiveTrait({
        name: "Token",
        description: "Token to use for fetching request tokens (if not using tokenUrl)",
        type: "string"
    })
], ArcGisImageServerCatalogItemTraits.prototype, "token", void 0);
__decorate([
    primitiveTrait({
        type: "number",
        name: "Maximum Refresh Intervals",
        description: "The maximum number of discrete times that can be created by a single " +
            "date range when layer in time-enabled."
    })
], ArcGisImageServerCatalogItemTraits.prototype, "maxRefreshIntervals", void 0);
__decorate([
    primitiveTrait({
        type: "number",
        name: "WKID",
        description: "The well-known ID of the spatial reference of the image server. Only Web Mercator (102100 or 102113 or 3857) and WGS84 (4326) are supported."
    })
], ArcGisImageServerCatalogItemTraits.prototype, "wkid", void 0);
__decorate([
    primitiveTrait({
        name: "Use Pre-Cached Tiles",
        description: "If true, the server's pre-cached tiles are used. If false, then the ImageServer exportImage endpoint will be used. This will default to false if parameters (including time) have been specified, otherwise it will default to true if a server supports pre-cached tiles.",
        type: "boolean"
    })
], ArcGisImageServerCatalogItemTraits.prototype, "usePreCachedTiles", void 0);
__decorate([
    primitiveArrayTrait({
        name: "Band IDs",
        description: "The band IDs to use when requesting images.",
        type: "number"
    })
], ArcGisImageServerCatalogItemTraits.prototype, "bandIds", void 0);
__decorate([
    objectTrait({
        name: "Rendering Rule",
        description: 'The rendering rule to apply to the image service. This must be a JSON object - for example `{"rasterFunction": "RFTAspectColor"}`. Note `allowRasterFunction` must be true for this to be applied.',
        type: ArcGisImageServerRenderingRule
    })
], ArcGisImageServerCatalogItemTraits.prototype, "renderingRule", void 0);
__decorate([
    primitiveTrait({
        name: "Allow Raster Function",
        description: "If true, then the renderingRule will be applied to the image service. If false, the renderingRule will be ignored. This will default to true if an image service supports raster functions.",
        type: "boolean"
    })
], ArcGisImageServerCatalogItemTraits.prototype, "allowRasterFunction", void 0);
__decorate([
    objectArrayTrait({
        type: ArcGisImageServerAvailableRasterFunctionTraits,
        name: "Available raster functions",
        description: "The available raster functions for the ImageServer. Defaults to all raster functions in the service if the server supports raster functions. Note: `allowRasterFunction` must be true. To set the default raster function, use the `renderingRule` property.",
        idProperty: "name"
    })
], ArcGisImageServerCatalogItemTraits.prototype, "availableRasterFunctions", void 0);
__decorate([
    primitiveTrait({
        name: "Disable raster functions selectors",
        description: "When true, disables the dimension selectors in the workbench. This will default to true if the server does not support raster functions.",
        type: "boolean"
    })
], ArcGisImageServerCatalogItemTraits.prototype, "disableRasterFunctionSelectors", void 0);
ArcGisImageServerCatalogItemTraits = __decorate([
    traitClass({
        example: {
            url: "https://sampleserver6.arcgisonline.com/arcgis/rest/services/CharlotteLAS/ImageServer",
            type: "esri-imageServer",
            name: "CharlotteLAS"
        }
    })
], ArcGisImageServerCatalogItemTraits);
export default ArcGisImageServerCatalogItemTraits;
