// terriajs/dist/Traits/TraitsClasses/AssImpCatalogItemTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import primitiveArrayTrait from "../Decorators/primitiveArrayTrait";
import primitiveTrait from "../Decorators/primitiveTrait";
import { traitClass } from "../Trait";
import mixTraits from "../mixTraits";
import GltfCatalogItemTraits from "./GltfCatalogItemTraits";
let AssImpCatalogItemTraits = class AssImpCatalogItemTraits extends mixTraits(GltfCatalogItemTraits) {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "urls", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
        Object.defineProperty(this, "baseUrl", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
    }
};
__decorate([
    primitiveArrayTrait({
        type: "string",
        name: "URLs",
        description: `An array of URLs`
    })
], AssImpCatalogItemTraits.prototype, "urls", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "Base URL",
        description: `The base URL that paths in the 3D model (eg textures) are relative to`
    })
], AssImpCatalogItemTraits.prototype, "baseUrl", void 0);
AssImpCatalogItemTraits = __decorate([
    traitClass({
        description: `Creates one catalog item from url that points to a 3d model.`,
        example: {
            type: "assimp",
            name: "assimp example",
            urls: [
                "https://raw.githubusercontent.com/kovacsv/assimpjs/main/examples/testfiles/cube_with_materials.obj",
                "https://raw.githubusercontent.com/kovacsv/assimpjs/main/examples/testfiles/cube_with_materials.mtl"
            ],
            baseUrl: "https://github.com/kovacsv/assimpjs/raw/main/examples/testfiles/",
            scale: 100,
            origin: {
                latitude: -42.8826,
                longitude: 147.3257,
                height: 100
            },
            rectangle: {
                west: 147.33,
                south: -42.9,
                east: 147.33,
                north: -42.87
            },
            id: "some unique id"
        }
    })
], AssImpCatalogItemTraits);
export default AssImpCatalogItemTraits;
