// terriajs/dist/Traits/TraitsClasses/Cesium3DTilesCatalogItemTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { traitClass } from "../Trait";
import mixTraits from "../mixTraits";
import CatalogMemberTraits from "./CatalogMemberTraits";
import Cesium3DTilesTraits from "./Cesium3dTilesTraits";
import FeatureInfoUrlTemplateTraits from "./FeatureInfoTraits";
import MappableTraits from "./MappableTraits";
import PlaceEditorTraits from "./PlaceEditorTraits";
import SearchableItemTraits from "./SearchableItemTraits";
import ShadowTraits from "./ShadowTraits";
import TransformationTraits from "./TransformationTraits";
import UrlTraits from "./UrlTraits";
let Cesium3DTilesCatalogItemTraits = class Cesium3DTilesCatalogItemTraits extends mixTraits(SearchableItemTraits, PlaceEditorTraits, TransformationTraits, FeatureInfoUrlTemplateTraits, MappableTraits, UrlTraits, CatalogMemberTraits, ShadowTraits, Cesium3DTilesTraits) {
};
Cesium3DTilesCatalogItemTraits = __decorate([
    traitClass({
        description: `Creates a 3d tiles item in the catalog from an ION Asset ID.
  <strong>Note:</strong> <i>Instead of specifying <b>ionAssetId</b> property, you can also provide a URL, for example, <code>"url": "https://storage.googleapis.com/vic-datasets-public/1ce41fe7-aed2-4ad3-be4d-c38b715ce9af/v1/tileset.json"</code>.</i>`,
        example: {
            type: "3d-tiles",
            ionAssetId: 69380,
            name: "CoM Melbourne 3D Photo Mesh",
            id: "some-unique-id"
        }
    })
], Cesium3DTilesCatalogItemTraits);
export default Cesium3DTilesCatalogItemTraits;
