// terriajs/dist/Traits/TraitsClasses/CompositeCatalogItemTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import CatalogMemberFactory from "../../Models/Catalog/CatalogMemberFactory";
import modelReferenceArrayTrait from "../Decorators/modelReferenceArrayTrait";
import mixTraits from "../mixTraits";
import { traitClass } from "../Trait";
import CatalogMemberTraits from "./CatalogMemberTraits";
import LegendOwnerTraits from "./LegendOwnerTraits";
import MappableTraits from "./MappableTraits";
let CompositeCatalogItemTraits = class CompositeCatalogItemTraits extends mixTraits(MappableTraits, CatalogMemberTraits, LegendOwnerTraits) {
    constructor() {
        super(...arguments);
        Object.defineProperty(this, "members", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: []
        });
    }
};
__decorate([
    modelReferenceArrayTrait({
        name: "Members",
        description: "The members of this composite.",
        factory: CatalogMemberFactory
    })
], CompositeCatalogItemTraits.prototype, "members", void 0);
CompositeCatalogItemTraits = __decorate([
    traitClass({
        description: `Creates one catalog item that combines multiple datasets.`,
        example: {
            type: "composite",
            name: "Combine Multiple Datasets",
            id: "some unique ID",
            members: [
                {
                    url: "https://portal.spatial.nsw.gov.au/server/rest/services/NSW_Administrative_Boundaries_Theme/MapServer/8",
                    id: "one unique ID",
                    type: "esri-mapServer"
                },
                {
                    url: "https://tiles.terria.io/static/auspost-locations.csv",
                    id: "another unique ID",
                    type: "csv"
                }
            ]
        }
    })
], CompositeCatalogItemTraits);
export default CompositeCatalogItemTraits;
