// terriajs/dist/Traits/TraitsClasses/GltfCatalogItemTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { traitClass } from "../Trait";
import mixTraits from "../mixTraits";
import AutoRefreshingTraits from "./AutoRefreshingTraits";
import CesiumIonTraits from "./CesiumIonTraits";
import GltfTraits from "./GltfTraits";
import PlaceEditorTraits from "./PlaceEditorTraits";
import UrlTraits from "./UrlTraits";
let GltfCatalogItemTraits = class GltfCatalogItemTraits extends mixTraits(UrlTraits, AutoRefreshingTraits, PlaceEditorTraits, GltfTraits, CesiumIonTraits) {
};
GltfCatalogItemTraits = __decorate([
    traitClass({
        description: `Creates one catalog item from url that points to a gltf file.`,
        example: {
            name: "gltf example",
            type: "gltf",
            url: "https://tiles.terria.io/terriajs-examples/czml/geelong_mini_demo_bin_linux_v1.0-draft_x64.gltf",
            origin: {
                longitude: 144.3569,
                latitude: -38.14688,
                height: -16
            },
            heightReference: "CLAMP_TO_TERRAIN",
            id: "some unique id"
        }
    })
], GltfCatalogItemTraits);
export default GltfCatalogItemTraits;
