// terriajs/dist/Traits/TraitsClasses/I3SCatalogItemTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { traitClass } from "../Trait";
import mixTraits from "../mixTraits";
import I3STraits from "./I3STraits";
import Cesium3DTilesCatalogItemTraits from "./Cesium3DTilesCatalogItemTraits";
let I3SCatalogItemTraits = class I3SCatalogItemTraits extends mixTraits(Cesium3DTilesCatalogItemTraits, I3STraits) {
};
I3SCatalogItemTraits = __decorate([
    traitClass({
        description: `Creates an I3S item in the catalog from an slpk.`,
        example: {
            type: "I3S",
            name: "CoM Melbourne 3D Photo Mesh",
            id: "some-unique-id"
        }
    })
], I3SCatalogItemTraits);
export default I3SCatalogItemTraits;
