// terriajs/dist/Traits/TraitsClasses/SdmxCatalogGroupTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import CatalogMemberTraits from "./CatalogMemberTraits";
import GroupTraits from "./GroupTraits";
import mixTraits from "../mixTraits";
import SdmxCommonTraits from "./SdmxCommonTraits";
import UrlTraits from "./UrlTraits";
import LegendOwnerTraits from "./LegendOwnerTraits";
import { traitClass } from "../Trait";
let SdmxCatalogGroupTraits = class SdmxCatalogGroupTraits extends mixTraits(SdmxCommonTraits, UrlTraits, CatalogMemberTraits, LegendOwnerTraits, GroupTraits) {
};
SdmxCatalogGroupTraits = __decorate([
    traitClass({
        description: `Creates one catalog group from url that points to an sdmx service.`,
        example: {
            id: "some unique id",
            type: "sdmx-group",
            name: "sdmx-group example",
            url: "https://api.data.abs.gov.au",
            mergeGroupsByName: true
        }
    })
], SdmxCatalogGroupTraits);
export default SdmxCatalogGroupTraits;
