// terriajs/dist/Traits/TraitsClasses/ShapefileCatalogItemTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { GeoJsonTraits } from "./GeoJsonTraits";
import mixTraits from "../mixTraits";
import CatalogMemberTraits from "./CatalogMemberTraits";
import { traitClass } from "../Trait";
let ShapefileCatalogItemTraits = class ShapefileCatalogItemTraits extends mixTraits(GeoJsonTraits, CatalogMemberTraits) {
};
ShapefileCatalogItemTraits = __decorate([
    traitClass({
        description: `Creates one catalog item from url that points to a zipped shapefile.`,
        example: {
            type: "shp",
            name: "shp (shapefile) example",
            url: "https://tiles.terria.io/terriajs-examples/shp/airports.zip",
            id: "some unique id"
        }
    })
], ShapefileCatalogItemTraits);
export default ShapefileCatalogItemTraits;
