// terriajs/dist/Traits/TraitsClasses/ThreddsCatalogGroupTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import CatalogMemberTraits from "./CatalogMemberTraits";
import GroupTraits from "./GroupTraits";
import mixTraits from "../mixTraits";
import UrlTraits from "./UrlTraits";
import LegendOwnerTraits from "./LegendOwnerTraits";
import { traitClass } from "../Trait";
let ThreddsCatalogGroupTraits = class ThreddsCatalogGroupTraits extends mixTraits(GroupTraits, UrlTraits, CatalogMemberTraits, LegendOwnerTraits) {
};
ThreddsCatalogGroupTraits = __decorate([
    traitClass({
        description: `Creates one catalog group from url that points to a thredds service.`,
        example: {
            type: "thredds-group",
            url: "http://dapds00.nci.org.au/thredds/catalog/rr9/ASCAT/ASCAT_v1-0_soil-moisture_daily_0-05deg_2007-2011/00000000/catalog.xml",
            name: "thredds example",
            id: "some unique id"
        }
    })
], ThreddsCatalogGroupTraits);
export default ThreddsCatalogGroupTraits;
