// terriajs/dist/Traits/TraitsClasses/WebFeatureServiceCatalogGroupTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import CatalogMemberTraits from "./CatalogMemberTraits";
import GetCapabilitiesTraits from "./GetCapabilitiesTraits";
import GroupTraits from "./GroupTraits";
import mixTraits from "../mixTraits";
import UrlTraits from "./UrlTraits";
import LegendOwnerTraits from "./LegendOwnerTraits";
import { traitClass } from "../Trait";
let WebFeatureServiceCatalogGroupTraits = class WebFeatureServiceCatalogGroupTraits extends mixTraits(GetCapabilitiesTraits, GroupTraits, UrlTraits, CatalogMemberTraits, LegendOwnerTraits) {
};
WebFeatureServiceCatalogGroupTraits = __decorate([
    traitClass({
        description: `Creates a group of all layers in the catalog from a url that points to a WFS service.`,
        example: {
            type: "wfs-group",
            name: "wfs-group example",
            url: "https://warehouse.ausseabed.gov.au/geoserver/ows",
            id: "some unique id for wfs-group example"
        }
    })
], WebFeatureServiceCatalogGroupTraits);
export default WebFeatureServiceCatalogGroupTraits;
